images/virtualization-artifact/pkg/controller/vmclass/internal/discovery.go

```diff
@@ -0,0 +1,151 @@
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/deckhouse/virtualization-controller/pkg/controller/conditions"
+	"github.com/deckhouse/virtualization-controller/pkg/controller/vmclass/internal/state"
+	virtv2 "github.com/deckhouse/virtualization/api/core/v1alpha2"
+	"github.com/deckhouse/virtualization/api/core/v1alpha2/vmclasscondition"
+)
+
+const nameDiscoveryHandler = "DiscoveryHandler"
+
+func NewDiscoveryHandler() *DiscoveryHandler {
+	return &DiscoveryHandler{}
+}
+
+type DiscoveryHandler struct{}
+
+func (h *DiscoveryHandler) Handle(ctx context.Context, s state.VirtualMachineClassState) (reconcile.Result, error) {
+	if s.VirtualMachineClass().IsEmpty() {
+		return reconcile.Result{}, nil
+	}
+	current := s.VirtualMachineClass().Current()
+	changed := s.VirtualMachineClass().Changed()
+
+	if updated := addAllUnknown(changed, vmclasscondition.TypeDiscovered.String()); updated {
+		return reconcile.Result{Requeue: true}, nil
+	}
+
+	if isDeletion(current) {
+		return reconcile.Result{}, nil
+	}
+
+	t := current.Spec.CPU.Type
+	mgr := conditions.NewManager(changed.Status.Conditions)
+	if t == virtv2.CPUTypeHostPassthrough || t == virtv2.CPUTypeHost {
+		mgr.Update(conditions.NewConditionBuilder2(vmclasscondition.TypeDiscovered).
+			Generation(current.GetGeneration()).
+			Message(fmt.Sprintf("Discovery available only cpu.type %q", virtv2.CPUTypeDiscovery)).
+			Reason2(vmclasscondition.ReasonDiscoverySkip).
+			Status(metav1.ConditionFalse).Condition())
+		changed.Status.Conditions = mgr.Generate()
+		return reconcile.Result{}, nil
+	}
+
+	nodes, err := s.Nodes(ctx)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	availableNodes := make([]string, len(nodes))
+	for i, n := range nodes {
+		availableNodes[i] = n.GetName()
+	}
+
+	var featuresEnabled []string
+	switch t {
+	case virtv2.CPUTypeDiscovery:
+		featuresEnabled = h.discoveryCommonFeatures(nodes)
+	case virtv2.CPUTypeFeatures:
+		featuresEnabled = current.Spec.CPU.Features
+	}
+
+	var featuresNotEnabled []string
+	if t == virtv2.CPUTypeDiscovery || t == virtv2.CPUTypeFeatures {
+		selectedNodes, err := s.NodesByVMNodeSelector(ctx)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+		commonFeatures := h.discoveryCommonFeatures(selectedNodes)
+		for _, cf := range commonFeatures {
+			if !slices.Contains(featuresEnabled, cf) {
+				featuresNotEnabled = append(featuresNotEnabled, cf)
+			}
+		}
+	}
+	cb := conditions.NewConditionBuilder2(vmclasscondition.TypeDiscovered).Generation(current.GetGeneration())
+	switch t {
+	case virtv2.CPUTypeDiscovery:
+		if len(featuresEnabled) > 0 {
+			cb.Message("").Reason2(vmclasscondition.ReasonDiscoverySucceeded).Status(metav1.ConditionTrue)
+			break
+		}
+		cb.Message("Detection nodes have no features.").
+			Reason2(vmclasscondition.ReasonDiscoveryFailed).
+			Status(metav1.ConditionFalse)
+	default:
+		cb.Message(fmt.Sprintf("Discovery available only cpu.type %q", virtv2.CPUTypeDiscovery)).
```

Comment:
```suggestion
		cb.Message(fmt.Sprintf("Discovery not needed for cpu.type %q", t)).
```

Message(fmt.Sprintf("Discovery not needed for cpu.type %q", t)).